### ISSUE TITLE
Time to connect is usually one of 3 values (banding)

### ISSUE THREAD

**Issue by Thinqer77** [knolleary/pubsubclient#621]
Hi,

I've recently set up some ESP8266 based temperature sensors around my house that connect to an MQTT broker (running on a Synology box). My wireless network is Ubiquti UniFi kit and I have 4 APs dotted around the house giving me pretty good coverage. The temperature sensors tend to connect to whichever AP is closest to them. Every 3 minutes they wake up, connect to wifi, connect to MQTT and publish the temperature along with some logging info like RSSI, voltage, how long it took to connect to the wifi and how long it took to connect to the MQTT broker.

I've noticed that all of the sensors have distinct 'bands' into which the MQTT connection times fall. These are consistent regardless of which AP they connect to. The bands are around 2.75s, around 800ms and 60ms - the following image shows what I mean (the charts on the right):

![Temperature sensor metrics](https://user-images.githubusercontent.com/16841861/57575698-740cb280-7447-11e9-9ed3-9f4313030230.JPG)

Just wondering if anyone else has seen this sort of behaviour before and whether or not it's expected when using this library? My aim is to get the time to connect as low as possible so the devices is awake for as little time as possible to conserve battery.

Any suggestions greatly appreciated - happy to post code if it helps.

Ben

**Comment by TD-er** [comment]
I guess this is not really PubSubClient related.

To my knowledge, the connection time of the ESP8266 (and others) greatly depends on the following (first is always faster):
- Static IP vs dynamic IP
- Fixed channel vs dynamic channel
- Connect based on IP or based on DNS

On top of that is some improvement in response time noticable when communication between devices is on the same WiFi accesspoint, compared to connection via different AP's.

**Comment by Thinqer77** [comment]
Yeah, I'd tend to agree regarding connection times, but my timing is done **after** the ESP8266 is connected to the WiFi network; the green chart shows how long it took to connect to WiFi, the purple chart is solely the time it takes to connect to the MQTT broker and as you can see it has some pretty consistent time banding across all my devices (some of which connect to different APs on my network).

In terms of your suggestions for speeding up connection, I use static IP on a fixed channel and connect based on IP.

The broker is running on a Synology box (wired into the network) so the problem could lie there I suppose and I should really test that but thought I'd try here first to see if anyone else had noticed any banding like I have.

**Comment by TD-er** [comment]
Maybe it's a multiple of some timeout value?
